contracts/protocol/facets/ProtocolInitializationFacet.sol

```diff
@@ -0,0 +1,106 @@
+import { DiamondLib } from "../../diamond/DiamondLib.sol";
+
+/**
+ * @title IBosonProtocolInitializationHandler
+ *
+ * @notice Handle initializion of new versions after 2.1.0.
+ *
+ */
+contract ProtocolInitializationFacet is IBosonProtocolInitializationHandler, ProtocolBase {
+    /**
+     * @notice Modifier to protect initializer function from being invoked twice for a given version.
+     */
+    modifier onlyUnInitializedVersion(bytes32 _version) {
+        ProtocolLib.ProtocolStatus storage ps = protocolStatus();
+        require(!ps.initializedVersions[_version], ALREADY_INITIALIZED);
+        ps.initializedVersions[_version] = true;
+        _;
+    }
+
+    /**
+     * @notice Initializes the protocol after the deployment.
+     * This function is callable only once for each version
+     *
+     * @param _version - version of the protocol
+     * @param _addresses - array of facet addresses to call initialize methods
+     * @param _calldata -  array of facets initialize methods encoded as calldata
+     *                    _calldata order must match _addresses order
+     * @param _isUpgrade - flag to indicate whether this is first deployment or upgrade
+     * @param interfacesToRemove - array of interfaces to remove from the diamond
+     * @param interfacesToAdd - array of interfaces to add to the diamond
+     */
+    function initialize(
+        bytes32 _version,
+        address[] calldata _addresses,
+        bytes[] calldata _calldata,
+        bool _isUpgrade,
+        bytes4[] calldata interfacesToRemove,
+        bytes4[] calldata interfacesToAdd
+    ) external onlyUnInitializedVersion(_version) {
+        require(_version != bytes32(0), VERSION_MUST_BE_SET);
+        require(_addresses.length == _calldata.length, ADDRESSES_AND_CALLDATA_LENGTH_MISMATCH);
+
+        // Delegate call to initialize methods of facets declared in _addresses
+        for (uint256 i = 0; i < _addresses.length; i++) {
+            (bool success, bytes memory error) = _addresses[i].delegatecall(_calldata[i]);
+
+            // Handle result
+            if (!success) {
+                if (error.length > 0) {
+                    // bubble up the error
+                    revert(string(error));
+                } else {
+                    // Reverts with default message
+                    revert(PROTOCOL_INITIALIZATION_FAILED);
+                }
+            }
+        }
+
+        if (_isUpgrade) {
+            if (_version == bytes32("2.2.0")) {
+                initV2_2_0();
+            }
+        }
+
+        removeInterfaces(interfacesToRemove);
+        addInterfaces(interfacesToAdd);
+
+        ProtocolLib.ProtocolStatus storage status = protocolStatus();
+        status.version = _version;
+        emit ProtocolInitialized(string(abi.encodePacked(_version)));
+    }
+
+    /**
+     * @notice Initializes the version 2.2.0.
+     *
+     */
+    function initV2_2_0() internal {}
```

Comment:
why bother with this ? If the body is empty ? Is it meant to be illustrative?

Reply:
see @zajck comment:

> Beside what I commented directly in the code.
> 
> 1. `initV2_2_0` will not be empty, since the changes in #485 introduce a new config variable, which should be set in `initV2_2_0`.
>    But even if `initV2_2_0` remained empty, I would still keep it in the code, since it does not increase code size (compiler is smart enough). Once we have multiple `V2_x_y` it will be clearer what each upgrade does if we keep full list.
> 2. One thing I didn't comment in the code, but I think it might be sensible to do: we said that in future upgrades we'd check that version is increased only by one (i.e. v2.3.0 will require that existing version will be v2.2.x). Maybe in this first upgrade, it would make sense to check that version is 0.
> 3. [This is probably more relevant to upgrade-script PR]. In upgrade script we now remove supportedInterface 1 by 1. Maybe we could change that and this new facet would accept a list of supportedInterface to remove and do it during the initialization.

Reply:
Ok, makes sense, thanks. i will do a better job of following all of the chat :)